### PR TITLE
Added document type parameter to Storage method signatures and more tests

### PR DESCRIPTION
**Pull request by nikitag-ms**
* [-] Has `CHANGELOG.md` been updated?
* [+] Are tests passing locally?
* [+] Are the files formatted correctly?
* [+] Did you add unit tests?
* [+] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [+] Did you check UI tests on the sample app? They are not executed on CI.

## Description
- Added document type parameter to Storage method signatures to work around Java erasure
- Tests coverage is almost 100%, going to add a few more tests with the next commit to this PR

### CODE REVIEW

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 1]
```diff
@@ -162,52 +160,55 @@ protected String getLoggerTag() {
...
-                        callCosmosDbReadApi(tokenResult, documentId, result);
+                        callCosmosDbReadApi(tokenResult, documentId, documentType, result);
                     }
 
                     @Override
                     public void completeFuture(Exception e) {
                         completeFutureAndRemovePendingCall(e, result);
-                    }
-                });
+                    }});
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: is the format correct?

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 2]
```diff
@@ -162,52 +160,55 @@ protected String getLoggerTag() {
...
-                        completeFutureAndRemovePendingCall(Utils.<T>parseDocument(payload), result);
+                        completeFutureAndRemovePendingCall(Utils.<T>parseDocument(payload, documentType), result);
                     }
 
                     @Override
                     public void onCallFailed(Exception e) {
                         completeFutureAndRemovePendingCall(e, result);
-                    }
-                });
+                    }});
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: is this format correct?

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 3]
```diff
@@ -219,7 +220,7 @@ public void onCallFailed(Exception e) {
      * List (need optional signature to configure page size)
      * The document type (T) must be JSON deserializable
      */
-    public <T> AppCenterFuture<Documents<T>> list(String partition, Class<T> documentType) {
+    public static  <T> AppCenterFuture<Documents<T>> list(String partition, Class<T> documentType) {
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: extra space

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 4]
```diff
@@ -231,39 +232,38 @@ public void onCallFailed(Exception e) {
...
-
-                    @Override
-                    public void completeFuture(Exception e) {
-                        completeFutureAndRemovePendingCall(e, result);
-                    }
-                });
+            partition,
+            result,
+            new TokenExchange.TokenExchangeServiceCallback() {
+                @Override
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: extra line

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 5]
```diff
@@ -277,14 +277,13 @@ public void completeFuture(Exception e) {
...
-                        completeFutureAndRemovePendingCall(Utils.<T>parseDocument(payload), result);
+                        completeFutureAndRemovePendingCall(Utils.<T>parseDocument(payload, documentType), result);
                     }
 
                     @Override
                     public void onCallFailed(Exception e) {
                         completeFutureAndRemovePendingCall(e, result);
-                    }
-                });
+                    }});
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: same here...

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 6]
```diff
@@ -295,7 +294,7 @@ public void onCallFailed(Exception e) {
      * Replace a document
      * The document instance (T) must be JSON serializable
      */
-    public <T> AppCenterFuture<Document<T>> replace(String partition, String documentId, T document) {
+    public static  <T> AppCenterFuture<Document<T>> replace(String partition, String documentId, T document){
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: extra space

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 7]
```diff
@@ -310,24 +309,21 @@ public void onCallFailed(Exception e) {
...
-
-                    @Override
-                    public void completeFuture(Exception e) {
-                        completeFutureAndRemovePendingCall(e, result);
-                    }
-                });
+            partition,
+            result,
+            new TokenExchange.TokenExchangeServiceCallback() {
+                @Override
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: extra line

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 8]
```diff
@@ -310,24 +309,21 @@ public void onCallFailed(Exception e) {
         return getInstance().instanceDelete(partition, documentId);
     }
 
-    private synchronized AppCenterFuture<Document<Void>> instanceDelete(final String partition, final String documentId) {
-        AppCenterLog.debug(LOG_TAG, String.format("Delete started for document with id: %s", documentId));
+    private synchronized  AppCenterFuture<Document<Void>> instanceDelete(final String partition, final String documentId){
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: space after `(`

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java [hunk 9]
```diff
@@ -33,13 +33,16 @@ public Document(T document, String partition, String id) {
...
-        this.documentError = new DocumentError(exception);
+    public Document(T document, String partition, String id, String eTag, long timestamp) {
+        this(document, partition, id);
+        this.eTag = eTag;
+        this.timestamp = timestamp;
+        this.document = document;
+    }
+
+    public Document(Exception dError)
+    {
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java]: bracket place wrong

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 10]
```diff
@@ -42,8 +46,10 @@
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.verifyNoMoreInteractions;
 import static org.powermock.api.mockito.PowerMockito.whenNew;
 
+
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: extra line

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 11]
```diff
@@ -136,6 +156,28 @@ public void disablePersisted() {
...
     }
 
+    @Test
+    public void listEndToEnd() throws Exception {
+        HttpClientRetryer httpClient = mock(HttpClientRetryer.class);
+        whenNew(HttpClientRetryer.class).withAnyArguments().thenReturn(httpClient);
+        when(SharedPreferencesManager.getBoolean(STORAGE_ENABLED_KEY, true)).thenReturn(false);
+        Storage storage = Storage.getInstance();
+        Channel channel = start(storage);
+
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: extra line

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 12]
```diff
@@ -136,6 +156,28 @@ public void disablePersisted() {
...
+    }
+
+    @Test
+    public void replaceEndToEnd() throws Exception {
+        HttpClientRetryer httpClient = mock(HttpClientRetryer.class);
+        whenNew(HttpClientRetryer.class).withAnyArguments().thenReturn(httpClient);
+        when(SharedPreferencesManager.getBoolean(STORAGE_ENABLED_KEY, true)).thenReturn(false);
+        Storage storage = Storage.getInstance();
+        Channel channel = start(storage);
+
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: extra line

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 13]
```diff
@@ -136,6 +156,28 @@ public void disablePersisted() {
...
+        AppCenterFuture<Document<TestDocument>> doc = Storage.replace(PARTITION, DOCUMENT_ID, new TestDocument(TEST_FIELD_VALUE));
+    }
+
     @Test
     public void readEndToEnd() throws Exception {
         HttpClientRetryer httpClient = mock(HttpClientRetryer.class);
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: add comments /* When */ /* Call */ /* verify */

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 14]
```diff
@@ -180,7 +230,80 @@ public void readEndToEnd() throws Exception {
...
+    }
+
+    @Test
+    public void readFailTokenExchangeReturnsFailedTokenResultPayload() throws Exception {
+        HttpClientRetryer httpClient = mock(HttpClientRetryer.class);
+        whenNew(HttpClientRetryer.class).withAnyArguments().thenReturn(httpClient);
+        when(SharedPreferencesManager.getBoolean(STORAGE_ENABLED_KEY, true)).thenReturn(false);
+        Storage storage = Storage.getInstance();
+        Channel channel = start(storage);
+
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: extra line in multiple places... just just one more point out  :D

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 15]
```diff
@@ -267,6 +390,37 @@ public void buildAppCenterGetDbTokenBodyPayload() {
         // These constructors must be called even though these classes are not going to be instantiated.
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: the single line comment should be /* comments here */

**Comment by YiMysty** [comment]
Please check the comment in this work item to help to fix it :D
https://msmobilecenter.visualstudio.com/Mobile-Center/_workitems/edit/57919

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 16]
```diff
@@ -27,9 +27,7 @@
 import static com.microsoft.appcenter.http.DefaultHttpClient.METHOD_GET;
 import static com.microsoft.appcenter.http.DefaultHttpClient.METHOD_POST;
 import static com.microsoft.appcenter.http.HttpUtils.createHttpClient;
-import static com.microsoft.appcenter.storage.Constants.LOG_TAG;
-import static com.microsoft.appcenter.storage.Constants.SERVICE_NAME;
-import static com.microsoft.appcenter.storage.Constants.STORAGE_GROUP;
+import static com.microsoft.appcenter.storage.Constants.*;
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: Please don't do that and always commit with via Android Studio with the following options:

screen shot 2019-02-28 at 4 51 14 pm

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 17]
```diff
@@ -219,7 +220,7 @@ public void onCallFailed(Exception e) {
      * List (need optional signature to configure page size)
      * The document type (T) must be JSON deserializable
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: please end sentences with dots. Global comment.

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java [hunk 18]
```diff
@@ -11,14 +13,24 @@
...
-    public static synchronized <T> Document<T> parseDocument(String documentPayload) {
-        return Utils.sGson.fromJson(documentPayload, new TypeToken<Document<T>>() {
-        }.getType());
+    public static synchronized <T> Document<T> parseDocument(String cosmosDbPayload, Class<T> documentType) {
+        JsonParser parser = new JsonParser();
+        JsonObject obj = parser.parse(cosmosDbPayload).getAsJsonObject();
+        T document = sGson.fromJson(obj.get("document"), documentType);
+        return new Document<T>(
+                document,
+                obj.get("PartitionKey").getAsString(),
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Utils.java]: Introduce constants.

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java [hunk 19]
```diff
@@ -33,13 +33,16 @@ public Document(T document, String partition, String id) {
...
-    public Document(Exception exception) {
-        this.documentError = new DocumentError(exception);
+    public Document(T document, String partition, String id, String eTag, long timestamp) {
+        this(document, partition, id);
+        this.eTag = eTag;
+        this.timestamp = timestamp;
+        this.document = document;
+    }
+
+    public Document(Exception dError)
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java]: Rename dError to exception

Comment (YiMysty) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java]: I thought it's renamed in my previous pr... what happened :thinking:

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java (hunk 13 above)
Comment (nikitag-ms) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: These particular ones are for code coverage, basically, stubs

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java (hunk 14 above)
Comment (nikitag-ms) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: I'll fix in the product code, but is it ok to keep it as in tests? @guperrot, what do you think?

Comment (YiMysty) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: I think we should fix this when once we saw it... :)

Comment (guperrot) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: add comments what you mock or use the "if" "when" "then" test generic comments

File: apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 20]
```diff
@@ -23,11 +23,29 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
         /* TODO remove reflection once Storage published to jCenter. */
         try {
             final Class<?> storage = Class.forName("com.microsoft.appcenter.storage.Storage");
-            storage.getMethod("read", String.class, String.class).invoke(null, "User124", "3456");
-            storage.getMethod("delete", String.class, String.class).invoke(null, "User123c456q", "34567006");
-            storage.getMethod("create", String.class, String.class, Object.class).invoke(null, "User1235", "dfrer", new Test());
-        } catch (Exception ignore) {
+            //createDocument(storage);
```
Comment (guperrot) [inline apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: Cleanup this

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Constants.java [hunk 21]
```diff
@@ -9,26 +9,27 @@
 
     public static final String TOKEN_RESULT_SUCCEED = "Succeed";
 
+    public static final String DOCUMENT_FIELD_NAME = "document";
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Constants.java]: blank lines between fields

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java [hunk 22]
```diff
@@ -33,13 +34,15 @@ public Document(T document, String partition, String id) {
...
-    public Document(Exception exception) {
-        this.documentError = new DocumentError(exception);
+    public Document(T document, String partition, String id, String eTag, long timestamp) {
+        this(document, partition, id);
+        this.eTag = eTag;
+        this.timestamp = timestamp;
+        this.document = document;
+    }
+
+    public Document(Exception dError) {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java]: dError => exception

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/TestDocument.java [hunk 23]
```diff
@@ -0,0 +1,10 @@
+package com.microsoft.appcenter.storage;
+
+public class TestDocument {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/TestDocument.java]: you use that in a lot of places, can you reuse this class in tests?

File: apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java (hunk 20 above)
Comment (nikitag-ms) [reply apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: These are actually used. Uncommented.

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 24]
```diff
@@ -144,28 +186,36 @@ public void readEndToEnd() throws Exception {
         Storage storage = Storage.getInstance();
         Channel channel = start(storage);
 
-        AppCenterFuture<Document<TestDocument>> doc = Storage.read(PARTITION, DOCUMENT_ID);
+        AppCenterFuture<Document<TestDocument>> doc = Storage.read(PARTITION, DOCUMENT_ID, TestDocument.class);
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: I am talking about test cases like this... need to have the test setup... /**setup **/ /** Call **/ /**verify */

Comment (guperrot) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: Keep in mind that `/**` with 2 stars is only at class, method or field level (Javadoc).
Regular code comments is just `/*`

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/TestDocument.java (hunk 23 above)
Comment (nikitag-ms) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/TestDocument.java]: Spoke offline and figured it out.

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java [hunk 25]
```diff
@@ -1,25 +1,26 @@
...
 import com.google.gson.annotations.SerializedName;
+import com.microsoft.appcenter.storage.Constants;
 import com.microsoft.appcenter.storage.Utils;
 
 /**
  * A document coming back from CosmosDB.
  */
 public class Document<T> {
-    @SerializedName(value = "PartitionKey")
+    @SerializedName(value = Constants.PARTITION_KEY_FIELD_NAME)
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java]: blank line between class declaration an anything else

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 26]
```diff
@@ -180,7 +228,80 @@ public void readEndToEnd() throws Exception {
...
+                eq(METHOD_POST),
+                anyMapOf(String.class, String.class),
+                any(HttpClient.CallTemplate.class),
+                tokenExchangeServiceCallbackArgumentCaptor.capture());
+        TokenExchange.TokenExchangeServiceCallback tokenExchangeServiceCallback = tokenExchangeServiceCallbackArgumentCaptor.getValue();
+        assertNotNull(tokenExchangeServiceCallback);
+        String ExceptionMessage = "Call to token exchange failed for whatever reason";
+        tokenExchangeServiceCallback.onCallFailed(new Exception(ExceptionMessage));
+
+        /**
```
Comment (guperrot) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: Javadoc comments with 2 stars is for documentation only.
Inside methods it's only 1.

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 27]
```diff
@@ -180,7 +228,80 @@ public void readEndToEnd() throws Exception {
...
+                ArgumentCaptor.forClass(TokenExchange.TokenExchangeServiceCallback.class);
+        verify(httpClient).callAsync(
+                endsWith(TokenExchange.GET_TOKEN_PATH_FORMAT),
+                eq(METHOD_POST),
+                anyMapOf(String.class, String.class),
+                any(HttpClient.CallTemplate.class),
+                tokenExchangeServiceCallbackArgumentCaptor.capture());
+        TokenExchange.TokenExchangeServiceCallback tokenExchangeServiceCallback = tokenExchangeServiceCallbackArgumentCaptor.getValue();
+        assertNotNull(tokenExchangeServiceCallback);
+        String ExceptionMessage = "Call to token exchange failed for whatever reason";
```
Comment (guperrot) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: ExceptionMessage => exceptionMessage

File: apps/sasquatch/src/main/res/layout/activity_storage.xml [hunk 28]
```diff
@@ -8,5 +8,5 @@
     <TextView
         android:layout_width="match_parent"
         android:layout_height="wrap_content"
-        android:text="Hello World"/>
+        android:text="Nothing here but a lot happening in Logcat"/>
```
Comment (guperrot) [inline apps/sasquatch/src/main/res/layout/activity_storage.xml]: move to string resource like everything else in the test app

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/client/TokenExchange.java [hunk 29]
```diff
@@ -110,7 +111,14 @@ public void onBeforeCalling(URL url, Map<String, String> headers) {
         @Override
         public void onCallSucceeded(String payload, Map<String, String> headers) {
             final TokenResult tokenResult = parseTokenResult(payload);
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/client/TokenExchange.java]: We usually don't use final at local or parameter level if we don't have to.
We use it if we can at member field or static level.

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/exception/TokenExchangeException.java [hunk 30]
```diff
@@ -0,0 +1,7 @@
+package com.microsoft.appcenter.storage.exception;
+
+public class TokenExchangeException extends Exception {
+    public TokenExchangeException(String message) {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/exception/TokenExchangeException.java]: blank line missing here between class and the first thing it declares

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java [hunk 31]
```diff
@@ -110,5 +113,4 @@ public String toString() {
     // When caching is supported:
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/Document.java]: /* */

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java (hunk 10 above)
Comment (guperrot) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: yes remove line 52

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 32]
```diff
@@ -267,12 +388,43 @@ public void buildAppCenterGetDbTokenBodyPayload() {
...
+        Document<TestDocument> d =
+                Utils.parseDocument(cosmosDbDocumentResponsePayload, TestDocument.class);
+        assertEquals(DOCUMENT_ID, d.getId());
+        assertEquals(PARTITION, d.getPartition());
+        assertEquals(TEST_FIELD_VALUE, d.getDocument().test);
+    }
+
+    @Test
+    public void documentSerialization() {
+        assertEquals(
```
Comment (guperrot) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: in the next method you improved formatting by using local variable, this one would really benefit local variables too

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java [hunk 33]
```diff
@@ -267,12 +388,43 @@ public void buildAppCenterGetDbTokenBodyPayload() {
...
+        Map<String, String> headers = CosmosDb.generateHeaders(PARTITION, "token");
+        assertEquals(5, headers.size());
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void urlEncogingThrowsNonExistingEncoding() {
+        CosmosDb.urlEncode("a string to encode", "An encoding that doesn't exist");
     }
 
     class TestDocument {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: Found it! ;)

Comment (guperrot) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: (Use TestDocument from top level class you defined)

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/TestDocument.java (hunk 23 above)
Comment (guperrot) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/TestDocument.java]: https://github.com/Microsoft/AppCenter-SDK-Android/pull/970#discussion_r261454407

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java (hunk 24 above)
Comment (nikitag-ms) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: I decided to move set up to `setUp` method in the parent class. So the tests are less cluttered now. It seems like it's a bit more readable and these comments may not be as necessary as they were before.

File: sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java (hunk 33 above)
Comment (nikitag-ms) [reply sdk/appcenter-storage/src/test/java/com/microsoft/appcenter/storage/StorageTest.java]: Hm, that is some funny refactoring corner case there! Because I used `Refactor` to move it to the separate file. But for some reason it didn't erase the original one. Good eye!
